core/platform/auth/firebase_auth_services.py

```diff
@@ -0,0 +1,213 @@
+    except (ValueError, firebase_exceptions.FirebaseError) as e:
+        logging.error(e)
+        return None
+    else:
+        return claims.get('sub', None)
+
+
+def get_user_id_from_auth_id(auth_id):
+    """Returns the user ID associated with the given auth ID.
+
+    Args:
+        auth_id: str. The auth ID.
+
+    Returns:
+        str|None. The user ID associated with the given auth ID, or None if no
+        association exists.
+    """
+    model = auth_models.UserIdByFirebaseAuthIdModel.get(auth_id, strict=False)
+    return None if model is None else model.user_id
+
+
+def get_multi_user_ids_from_auth_ids(auth_ids):
+    """Returns the user IDs associated with the given auth IDs.
+
+    Args:
+        auth_ids: list(str). The auth IDs.
+
+    Returns:
+        list(str|None). The user IDs associated with each of the given auth IDs,
+        or None for associations which don't exist.
+    """
+    return [
+        None if model is None else model.user_id
+        for model in auth_models.UserIdByFirebaseAuthIdModel.get_multi(auth_ids)
+    ]
+
+
+def associate_auth_id_to_user_id(pair):
+    """Commits the association between auth ID and user ID.
+
+    Args:
+        pair: auth_domain.AuthIdUserIdPair. The association to commit.
+
+    Raises:
+        Exception. The auth ID is already associated with a user ID.
+    """
+    auth_id, user_id = pair
+
+    claimed_user_id = get_user_id_from_auth_id(auth_id)
+    if claimed_user_id is not None:
+        raise Exception(
+            'auth_id=%r is already mapped to user_id=%r' % (
+                auth_id, claimed_user_id))
+
+    mapping = (
+        auth_models.UserIdByFirebaseAuthIdModel(id=auth_id, user_id=user_id))
+    mapping.update_timestamps()
+    mapping.put()
+
+
+def associate_multi_auth_ids_to_user_ids(pairs):
+    """Commits the associations between auth IDs and user IDs.
+
+    Args:
+        pairs: list(auth_domain.AuthIdUserIdPair). The associations to
+            commit.
+
+    Raises:
+        Exception. One or more auth ID associations already exist.
+    """
+    # Turn list(pair) to pair(list): https://stackoverflow.com/a/7558990/4859885
+    auth_ids, user_ids = python_utils.ZIP(*pairs)
+
+    claimed_user_ids = get_multi_user_ids_from_auth_ids(auth_ids)
+    if any(user_id is not None for user_id in claimed_user_ids):
+        existing_associations = sorted(
+            'auth_id=%r, user_id=%r' % (auth_id, user_id)
+            for auth_id, user_id in python_utils.ZIP(auth_ids, claimed_user_ids)
+            if user_id is not None)
+        raise Exception(
+            'associations already exist for: %r' % (existing_associations,))
+
+    mappings = [
+        auth_models.UserIdByFirebaseAuthIdModel(id=auth_id, user_id=user_id)
+        for auth_id, user_id in python_utils.ZIP(auth_ids, user_ids)
+    ]
+    auth_models.UserIdByFirebaseAuthIdModel.update_timestamps_multi(mappings)
+    auth_models.UserIdByFirebaseAuthIdModel.put_multi(mappings)
```

Comment:
Does this need to be done transactionally against all other operations above? What if an association is made in the middle of this operation?

Reply:
Hmm, that's a good point yes they should be made in a transaction. Done.